File: lib/src/infrastructure/company/dto/company.dto.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:linko/src/infrastructure/company/dto/tag.dto.dart';

part 'company.dto.g.dart';

@JsonSerializable()
class CompanyDto extends Equatable {
  int? id;
  String name;
  int? userId;
  String? banner;
  String? phoneNumber;
  String? ownerPhoneNumber;
  String? deliveryTime;
  String? deliveryFee;
  String? localImage;
  String? url;
  bool isActive;
  List<String> tags;

  CompanyDto(
      {this.id,
      required this.name,
      required this.userId,
      this.banner,
      this.phoneNumber,
      this.ownerPhoneNumber,
      this.url,
      this.deliveryFee,
      this.localImage,
      this.deliveryTime,
      this.tags = const [],
      this.isActive = false});

  @override
  List<Object?> get props => [
        id,
        name,
        localImage,
        userId,
        tags,
        banner,
        phoneNumber,
        ownerPhoneNumber,
        url,
        deliveryFee,
        deliveryTime,
        isActive
      ];

  factory CompanyDto.fromJson(Map<String, dynamic> json) =>
      _$CompanyDtoFromJson(json);

  Map<String, dynamic> toJson() => _$CompanyDtoToJson(this);

  dynamic get(String propName) {
    switch (propName) {
      default:
        return name;
    }
  }

  static Future<List<CompanyDto>> loadCompanies(l) async {
    return List<CompanyDto>.from(l.map((model) => CompanyDto.fromJson(model)));
  }
}

File: src/modules/companies/domain/dtos/company.dto.ts
import { Company } from '../entities/company.entity';

export class CompanyDto {
  id: number;

  name: string;

  arabicName: string;

  userId: number;

  banner: string;

  url: string;

  phoneNumber: string;

  ownerPhoneNumber: string;

  deliveryTime: string;

  deliveryFee: string;

  isActive: boolean;

  badge: string;

  badgeColor: string;

  tags: string[];

  constructor(data: any) {
    if (data == null) return;
    if (data.name != null) this.name = data.name;
    if (data.arabicName != null) this.arabicName = data.arabicName;
    if (data.url != null) this.url = data.url;
    if (data.id != null) this.id = data.id;
    if (data.userId != null) this.userId = data.userId;
    if (data.banner != null) this.banner = data.banner;
    if (data.phoneNumber != null) this.phoneNumber = data.phoneNumber;
    if (data.ownerPhoneNumber != null)
      this.ownerPhoneNumber = data.ownerPhoneNumber;
    if (data.deliveryTime != null) this.deliveryTime = data.deliveryTime;
    if (data.deliveryFee != null) this.deliveryFee = data.deliveryFee;
    if (data.isActive != null) this.isActive = data.isActive;
    else data.isActive = false;
    if (data.badge != null) this.badge = data.badge;
    if (data.badgeColor != null) this.badgeColor = data.badgeColor;
    if (data.tags != null)
      this.tags =
        typeof data.tags == 'string' ? JSON.parse(data.tags) : data.tags;
  }

  static from(entity: Company): CompanyDto {
    const dto = new CompanyDto(null);
    dto.id = entity.id;
    dto.name = entity.name;
    dto.arabicName = entity.arabicName;
    dto.url = entity.url;
    dto.userId = entity.userId;
    dto.banner = entity.banner;
    dto.ownerPhoneNumber = entity.ownerPhoneNumber;
    dto.deliveryTime = entity.deliveryTime;
    dto.deliveryFee = entity.deliveryFee;
    dto.phoneNumber = entity.phoneNumber;
    dto.badge = entity.badge;
    dto.badgeColor = entity.badgeColor;
    dto.isActive = entity.isActive;
    return dto;
  }

  static toEntity(dto: CompanyDto): Company {
    const entity = new Company();
    entity.id = dto.id;
    entity.name = dto.name;
    entity.arabicName = dto.arabicName;
    entity.userId = dto.userId;
    entity.banner = dto.banner;
    entity.url = dto.url;
    entity.phoneNumber = dto.phoneNumber;
    entity.ownerPhoneNumber = dto.ownerPhoneNumber;
    entity.deliveryTime = dto.deliveryTime;
    entity.deliveryFee = dto.deliveryFee;
    entity.isActive = dto.isActive;
    entity.badge = dto.badge;
    entity.badgeColor = dto.badgeColor;
    if (entity.arabicName == null) entity.arabicName = entity.name;
    return entity;
  }
}
